Wrappers/Python/cil/io/TXRMTXMDataReader.py

```diff
@@ -100,46 +128,59 @@ def read(self):
-                    pixel_size=[d_pixel_size/1000,d_pixel_size/1000])\
-                .set_angles(angles, angle_unit=self.angle_unit)
-        self._ag.dimension_labels =  ['angle', 'vertical', 'horizontal']
+        if dist_source_center != dist_center_detector:
+            # Pixelsize loaded in metadata is really the voxel size in um.
+            # We can compute the effective detector pixel size as the geometric
+            # magnification times the voxel size.
+            d_pixel_size = ((dist_source_center+dist_center_detector)/dist_source_center)*metadata['pixel_size']
                 
+            self._ag = AcquisitionGeometry.create_Cone3D(
+                [0,-dist_source_center, 0] , [ 0, dist_center_detector, 0] \
+                ) \
+                    .set_panel([metadata['image_width'], metadata['image_height']],\
+                        pixel_size=[d_pixel_size/1000,d_pixel_size/1000])\
+                    .set_angles(angles[image_range],angle_unit=self.angle_unit)
+        else:
+            self._ag = AcquisitionGeometry.create_Parallel3D()\
+                        .set_panel([metadata['image_width'], metadata['image_height']])\
+                        .set_angles(angles[image_range],angle_unit=self.angle_unit)
+        self._ag.dimension_labels =  ['angle', 'vertical', 'horizontal']
+
+        ext = os.path.splitext(self.txrm_file)
+        if ext[1] == '.txrm':
+            # normalise data by flatfield
+            data = data / metadata['reference']
+
+            for num,k in enumerate(image_range):
+                data[num,:,:] = np.roll(data[num,:,:], \
+                    (int(metadata['x-shifts'][k]),int(metadata['y-shifts'][k])), \
+                    axis=(1,0))
+
+            acq_data = AcquisitionData(array=data, deep_copy=False, geometry=self._ag.copy(),\
+                suppress_warning=True)
+            self._metadata = metadata
+            return acq_data
+        else:
+            slices,width,height = data.shape
+            voxel_size = metadata["pixel_size"]
+            self._ig = ImageGeometry(voxel_num_x=height,
+                                    voxel_size_x=voxel_size,
+                                    voxel_num_y=width,
+                                    voxel_size_y=voxel_size,
+                                    voxel_num_z=slices,
+                                    voxel_size_z=voxel_size)
+            
+            ig_data = ImageData(array=data, deep_copy=False, geometry=self._ig)
+            self._metadata = metadata
+            return ig_data
 
-        acq_data = AcquisitionData(array=data, deep_copy=False, geometry=self._ag.copy(),\
-            suppress_warning=True)
-        self._metadata = metadata
-        return acq_data
     def load_projections(self):
         '''alias of read for backward compatibility'''
         return self.read()
-        
+
     def get_metadata(self):
```

Comment:
This seems to return the right metadata only if the data has been read?

Reply:
yes, I will change this in the new iteration.